UI: Panel

Amstutz opened the pull request:
Proposal for UI Panel Components, to be discussed in the next JF.

Amstutz committed:
UI: Panel, Declaration

Amstutz committed:
UI: Panel, Implementation

Amstutz committed:
UI: Panel, Templates

Amstutz committed:
UI: Panel, Tests

Amstutz committed:
UI: Panels, examples

Amstutz committed:
UI: Generic, Specification

Amstutz committed:
UI: Generic, Implementation

Amstutz committed:
UI: Generic, Tests

Amstutz committed:
UI: Generic, examples

klees commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,95 @@
+	 *      2: Standard Panels SHOULD be used in the Center Content as primary Container for grouping content of varying content.
+	 * ---
+	 * @param string $title
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
I think the CLASS1|CLASS2 notation could be used instead of mixed, so you could pull the part in the back to the front. You could also `use \ILIAS\UI\Component\Component` to get `Component[]|Component`.

klees replied:
Same in other places, btw.

klees commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,95 @@
+	 *   usage:
+	 *      1: Sub Panels MUST only be inside Standard Panels
+	 *   composition:
+	 *      1: Sub Panels MUST NOT contain Sub Panels or Standard Panels as content.
```
This can be checked easily in your implementation of `Panel\Sub`.

klees commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,95 @@
+	 *       Report Panels are predominantly used for displaying data. They may however comprise links or buttons.
+	 *   rivals:
+	 *      Standard Panels: >
+	 *        The Report Panels contains one Standard Panel containing the block panels used to structure information.
```
Is this correct? If i understand correctly, this conflicts with the actual parameters to report, which takes Subpanels.

Amstutz replied:
Legacy naming, good catch!

klees commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,95 @@
+	 *         is to be displayed alongside each other.
+	 *   interaction:
+	 *      1: Links MAY open new views.
+	 *      2: Buttons MAY trigger actions or inline editing.
```
Do these to lines really belong here?

Amstutz replied:
We talked about MAY rules before. They are of limited use and might just be part of the description. Point here is, that we like to make clear, for what Buttons and Links might be used in Reports. Do you think, they should be kicked out here?

klees replied:
No, this actually looked like a Copy-Paste-Error to me, as i didn't understand what these have to do with report panels =)

klees commented on src/UI/Component/Panel/Panel.php:
```diff
@@ -0,0 +1,21 @@
+	 * @param string $title Title of the Panel
+	 * @return \ILIAS\UI\Component\Panel\Panel
+	 */
+	public function withTitle($title);
```
Do we really want to change the title after construction?

klees commented on src/UI/Component/Panel/Report.php:
```diff
@@ -0,0 +1,32 @@
+	 * @param string $title Title of the Report
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function withTitle($title);
```
No need to duplicate this, its in Panel already.

klees commented on src/UI/Component/Panel/Report.php:
```diff
@@ -0,0 +1,32 @@
+	/**
+	 * @return string $title Title of the Report
+	 */
+	public function getTitle();
```
No need to duplicate this, its in Panel already.

klees commented on src/UI/Component/Panel/Report.php:
```diff
@@ -0,0 +1,32 @@
+	 * @param \ILIAS\UI\Component\Panel\Sub[] $sub_panels Sub Panels used to structure the report.
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function withSubPanels($sub_panels);
```
Do we really want to change this after construction?

Amstutz committed:
Fixed naming problem block --> sub

klees commented on src/UI/Component/Panel/Standard.php:
```diff
@@ -0,0 +1,21 @@
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
+	 * @return \ILIAS\UI\Component\Panel\Standard
+	 */
+	public function withContent($content);
```
Do we really want to change this after construction?

klees commented on src/UI/Component/Panel/Sub.php:
```diff
@@ -0,0 +1,34 @@
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
+	 * @return \ILIAS\UI\Component\Panel\Sub
+	 */
+	public function withContent($content);
```
Do we really want to change this after construction?

klees commented on src/UI/Component/Panel/Sub.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * This describes a Sub Panel.
+ */
+interface Sub extends Panel {
```
You could introduce another interface here that bundles the content-part of Sub and Standard.

Amstutz replied:
Is the rule each component should represent an interface only one way or two way? From your comment I gather it is one way. What name would you chose for this interface?

klees replied:
I'm not sure, i understand what you are asking. You could introduce `interface PanelWithContent` (proper naming pending) that has the `*Content` methods. This would deduplicate the `*Content` methods in Sub and Panel.

Amstutz replied:
Yes I could. What I am asking is: Do we need a Component named "Panel->WithContent" if we introduce this interface. I hope not.

lechimp-p replied:
I don't think we need that component.

klees commented on src/UI/Factory.php:
```diff
@@ -253,4 +253,22 @@ public function listing();
+	 *
+	 * rules:
+	 *   wording:
+	 *      1: Panels MUST contain a heading.
```
Maybe change to "Panels MUST have a title." as this is the parameter we actually supply to the panel constructors?

klees commented on src/UI/Implementation/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,34 @@
+use ILIAS\UI\Component\Panel as P;
+
+/**
+ * Class Factory
```
Just for curiosity: Is this a line PHPStorm always creates? I stumbled upon it very often already, as i consider this non-informative.

Amstutz replied:
:-) 

It depends on how one works. In my case PHP Storms complains if a class does not provide a description by Doc Comment. An easy fix for this complaint is to auto generate one. This is kind of a mental hack since the problem is not visible anymore. You are obviously right by pointing out that the issue of not having a description is actually still there.

klees commented on src/UI/Implementation/Component/Panel/Panel.php:
```diff
@@ -0,0 +1,82 @@
+	/**
+	 * @inheritdoc
+	 */
+	public function withContent($content){
```
Implementing this here is not a real problem, but this also hints at the fact, that there might be another interface for Panels with any content.

klees commented on tests/UI/Component/Panel/PanelTest.php:
```diff
@@ -0,0 +1,241 @@
+		$sub = $sub->withCard($card);
+
+		$report = $fp->report("Title",$sub);
+		$html = new DOMDocument();
```
This pattern looks nice. Maybe we could abstract it a little and promote it to the base class for tests?

klees replied:
Like "assertHTMLEquals" or something similar?

Amstutz replied:
I stumbled upon multiple issues while testing rather nested html output from our beloved template engine. Wrapping it with a DOMDocument helped reducing some of the problems.

+1 for assertHTMLEquals

klees replied:
Who is going to implement it, me or you, or...?

Amstutz replied:
Whoever blinks first. 
...
Oke, ill do it.

klees commented:
Will have another look on monday, not focused enough atm.

Amstutz committed:
Removed redundant methods

Amstutz committed:
Sugered doc comments

Amstutz committed:
changed wording in rules

Amstutz committed:
necessary fixes for report if panel used as base for tests to pass.

Amstutz committed:
Promoted testing logic to assertHTMLEquals

lechimp-p commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,96 @@
+	 *       Report Panels are predominantly used for displaying data. They may however comprise links or buttons.
+	 *   rivals:
+	 *      Standard Panels: >
+	 *        The Report Panels contains one Standard Panel containing the sub panels used to structure information.
```
This seems to contradict the signature of the factory method. The sig states, that the report panel contains sub panels.

Amstutz replied:
I slightly changed the description.

klees commented on src/UI/Component/Panel/Factory.php:
```diff
@@ -0,0 +1,96 @@
+	 * @param \ILIAS\UI\Component\Panel\Sub[] $sub_panels
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function report($title,$sub_panels);
```
Typehint on `array $sub_panels` possible.

klees commented on src/UI/Implementation/Component/Panel/Panel.php:
```diff
@@ -0,0 +1,80 @@
+	protected  $title;
+
+	/**
+	 * @var mixed content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
This is always Component[] by __construct.

Amstutz replied:
No it is not, it is Component[]|Component

klees commented on src/UI/Implementation/Component/Panel/Panel.php:
```diff
@@ -0,0 +1,80 @@
+
+	/**
+	 * @param string $title
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
`Component[]|Component`

klees commented on src/UI/Implementation/Component/Panel/Renderer.php:
```diff
@@ -0,0 +1,121 @@
+	 */
+	protected function render_standard(Component\Panel\Standard $component, RendererInterface $default_renderer)
+	{
+		$tpl = $this->getTemplate("tpl.panel.html", true, true);
```
Use a separate template for standard panel and report panel please. This would allow our designers to change them independently.

Amstutz replied:
The essence of a report panel is, that it consists of a panel and subpanel. I think it to be redundant here to generate an exact copy of the template just for the reason your proposing. In the tech meeting on thursday, we agreed that we should not set dom elements classes with variables. however we should also look to not generate too much redundancy in templates as alex pointed out.

It is possible that requirements for panels change in future, if so, we can think about splitting the templates.

klees commented on src/UI/Implementation/Component/Panel/Report.php:
```diff
@@ -0,0 +1,55 @@
+		$this->checkStringArg("title",$title);
+		$sub_panels= $this->toArray($sub_panels);
+		$types = [C\Panel\Sub::class];
+		$this->checkArgListElements("content", $sub_panels, $types);
```
This checks the `sub_panels` argument, not `content`.

klees commented on src/UI/Implementation/Component/Panel/Report.php:
```diff
@@ -0,0 +1,55 @@
+	 * @param string $title
+	 * @param \ILIAS\UI\Component\Panel\Sub[]
+	 */
+	public function __construct($title,$sub_panels) {
```
The `Report` class extends `Panel`, but doesn't invoke its constructor and thus leaves the inherited property `content` uninitialized, which makes `Panel::getContent` promise something it can't fullfill. Overall it looks to me as if `Report` is basically `Panel` with an additional constraints on the content. Maybe we should model it just like that.

Amstutz replied:
Currently we are kind of sitting between fragmented proposals. It would be helpful for me if you look first at the complete picture before proposing single fragments. Otherwise it is hard for me to work with your suggestions. I end up writing improvements for interface/templates/implementation/tests/examples 3 times.

My current solution is not good (on a structural level even buggy), however it is somewhat a result of fragmented proposals. WithContent was moved around quite a bit during the discussion.

From your current solution I gather that you propose the following:
We have an one interface panel. From my understanding of your proposal, the withTitle and withContent part. All panels would inherit from a base class panel implementing this interface, so no extra interface would be needed, right?

--> I can live with that, note however that if done as proposed, we can not type hint anymore on Panel\Sub in the withContent function of panel.

klees commented on src/UI/Component/Panel/Sub.php:
```diff
@@ -0,0 +1,34 @@
+	 * @param \ILIAS\UI\Component\Card\Card $card
+	 * @return Sub
+	 */
+	public function withCard($card);
```
Typehint on `Card` please.

klees commented on src/UI/Implementation/Component/Panel/Sub.php:
```diff
@@ -0,0 +1,40 @@
+	 * @inheritdoc
+	 */
+	public function withCard($card){
+		$this->checkArgInstanceOf("Card",$card,C\Card\Card::class);
```
If we typehint we don't need that call to `checkArgInstanceOf`.

klees commented on src/UI/Implementation/Component/Panel/Sub.php:
```diff
@@ -0,0 +1,40 @@
+
+	/**
+	 * Card to be displayed on the right of the Sub Panel
+	 * @var \ILIAS\UI\Component\Card\Card
```
`Card|null` please.

klees commented:
Maybe we should introduce scrutinizer for the src-directory to help us with reviews.

klees commented on tests/UI/Base.php:
```diff
@@ -92,4 +93,24 @@ public function getDefaultRenderer() {
+		$expected->formatOutput = true;
+		$expected->preserveWhiteSpace = false;
+
+		$html->loadXML($this->normalizeHTML($html_as_string));
```
Do we still need my homemade `normalizeHTML` or does `DOMDocument` do something similar for us?

Amstutz replied:
Currently we do to remove the new line breaks.

klees commented on tests/UI/Component/Listing/ListingTest.php:
```diff
@@ -103,26 +103,6 @@ public function test_render_ordered_listing() {
-			$this->assertEquals($e->getCode(), E_RECOVERABLE_ERROR);
-		}
-	}
-
```
Why did you remove that?

Amstutz replied:
Due to your recommendations we type hint on array for lists here. This results in an error in PHP7 (maybe also older versions) if string is given due to the type hint on PHP level. I did regard this test as superfluous since it makes limited sense to tests php type hinting capability. 

In this commit this is mainly here, that the tests run again.

klees replied:
Oka!

matthiaskunkel commented:
Jour Fixe, August 15, 2016: We highly appreciate this PR and please Timon to merge it into trunk.

Amstutz committed:
Renamed component and removed withContent

Amstutz committed:
Clarification of legacy component rule.

Amstutz committed:
Several changed to comply with suggestions given in feedback. Main changes:
--> Removed withContent and withTitle methods
--> Changed the interface of report to comply with the other two panels
--> Created seperate template for report

Amstutz committed:
Slightly changed wording in description

Amstutz committed:
Fixed examples to comply with new naming of legacy (aka generic) component.

Amstutz committed:
Removed withContent/withTitle tests --> all 41 panel tests are passing.

Amstutz committed:
Merge pull request #259 from Amstutz/UI_Generic

UI: Generic

Amstutz committed:
Merge branch 'trunk' into UI_Panel to remove merge conflicts with UI: Legacy (factories)

* trunk: (128 commits)
  added news update
  fixed bug 18794: Update from v4.3.4 breaks
  Added Forum Features: http://www.ilias.de/docu/goto.php?target=wiki_1357_Create_drafts_of_forum_postings, http://www.ilias.de/docu/goto.php?target=wiki_1357_Download_of_multiple_attachments_as_zip-file, http://www.ilias.de/docu/goto.php?target=wiki_1357_Forum_notifications_with_file_attachments, http://www.ilias.de/docu/goto.php?target=wiki_1357_Sorting_of_attachments_in_alphabetical_order
  #18805 - folder: download by anonymous
  Clarification of legacy component rule.
  Renamed component and removed withContent
  advanced editing: non static functions should not start with _
  fixed online status news creation
  test news item
  [FIX] 18819 missing require
  [FIX] replaced FETCHMODE_OBJECT with ilDBConstants::FETCHMODE_OBJECT
  copage: use geshi
  bookmarks: fixed template calls
  added missing static statement
  fixed constructors, and one signature
  copage: added editing info message for content includes
  media pool/copage: fixed style includes
  portfolio: content snippets in portfolio pages
  competences: historic dates/hide skills in personal competence view
  plugin: standalone import/export
  ...

Conflicts:
	src/UI/Factory.php
	src/UI/Implementation/Factory.php
	tests/UI/Base.php
